File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        
        ListNode * slow = head;
        ListNode * fast = head;
        while(fast!=NULL && fast->next!=NULL)
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode * mid = slow->next;
        slow->next = NULL;
        ListNode * prev = NULL;
        while(mid!=NULL)
        {
            ListNode *temp = mid->next;
            mid->next = prev;
            prev = mid;
            mid = temp;
        }

        ListNode *l1 = head;
        ListNode *l2 = prev;
        while(l1 && l2)
        {
            ListNode *l1next = l1->next;
            ListNode *l2next = l2->next;
            l1->next = l2;
            l2->next = l1next;
            l1 = l1next;
            l2 = l2next;
        }
        


        
    }
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head is None:
            return None 
        slow = head 
        fast = head 
        while fast and fast.next:
            slow = slow.next 
            fast = fast.next.next
        mid = slow
        l2 = mid.next 
        mid.next = None 
        prev = None 
        while l2:
            temp = l2.next 
            l2.next = prev 
            prev = l2
            l2 = temp 
        l2 = prev 
        l1 = head 
        while l1 and l2:
            l1next = l1.next 
            l2next = l2.next 
            l1.next = l2 
            l2.next = l1next 
            l1 = l1next 
            l2 = l2next
        return head
        